### DISCUSSION TITLE
Distinguish programmatically triggered navigation actions from user triggered ones

### DISCUSSION THREAD

**Discussion by asadwan** [react-navigation/react-navigation#12062] [Q&A]
Is there a way to distinguish programmatically triggered navigation actions from user triggered ones (e.g. Android back button or iOS back gesture)?

When using the `usePreventRemove` hook, I'd like to prevent remove only for user actions but it seems there's no way to distinguish between the two.

**Comment by satya164** [comment]
Triggering android back button or swipe gesture are also a user actions. There's no distinction.

If you want to distinguish specific actions triggered in your own components, then pass additional data in the actions/params that you can see in the action object.

**Reply by asadwan** [reply]
Thanks for your answer!

> Triggering android back button or swipe gesture are also a user actions. There's no distinction.

Sorry, I wasn't clear enough. What I meant is to distinguish between actions I trigger myself (push a new screen) or actions handled automatically by React Navigation (e.g. pop a screen from the stack in response to user pressing Android back button) 

> If you want to distinguish specific actions triggered in your own components, then pass additional data in the actions/params that you can see in the action object.

That's what I'm currently doing but the downside of it is that I lose type safety using `navigation.dispatch`. Is there a way to do it without losing type safety?

**Reply by satya164** [reply]
@asadwan you can pass that info in `params` to keep the type safety - as params are typechecked.

**Reply by asadwan** [reply]
In some cases, there are no `params`. For example, on a form screen:

1. If the form is dirty and the user tries to go back:
   - Prevent removal
   - Display confirmation alert
   - If user confirms, pop the screen
2. If the user submits the form:
   - Pop the screen

Both actions are `POP` without `params`. I use `navigation.dispatch` in 2 to add that info to the action's payload.
